src/streaming/ManifestUpdater.js

```diff
@@ -132,18 +134,67 @@ function ManifestUpdater() {
         }
     }
 
-    function refreshManifest() {
+    function refreshManifest(ignorePatch = false) {
         isUpdating = true;
         const manifest = manifestModel.getValue();
+
+        // default to the original url in the manifest
         let url = manifest.url;
+
+        // Check for PatchLocation and Location alternatives
+        const patchLocation = adapter.getPatchLocation(manifest);
         const location = adapter.getLocation(manifest);
-        if (location) {
+        if (patchLocation && !ignorePatch) {
+            url = patchLocation;
+        } else if (location) {
             url = location;
         }
+
+        // if one of the alternatives was relative, convert to absolute
+        if (urlUtils.isRelative(url)) {
+            url = urlUtils.resolve(url, manifest.url);
+        }
+
         manifestLoader.load(url);
     }
 
     function update(manifest) {
+        if (!manifest) {
+            // successful update with no content implies existing manifest remains valid
+            manifest = manifestModel.getValue();
+
+            // override load time to avoid invalid latency tracking
+            manifest.loadedTime = new Date();
+        } else if (adapter.getIsPatch(manifest)) {
+            // with patches the in-memory manifest is our base
+            let patch = manifest;
+            manifest = manifestModel.getValue();
+
+            // check for patch validity
+            let isPatchValid = adapter.isPatchValid(manifest, patch);
+            let patchSuccessful = isPatchValid;
+
+            if (isPatchValid) {
+                // grab publish time before update
+                let publishTime = adapter.getPublishTime(manifest);
+
+                // apply validated patch to manifest
+                patchSuccessful = adapter.applyPatchToManifest(manifest, patch);
+
+                // get the updated publish time
+                let updatedPublishTime = adapter.getPublishTime(manifest);
+
+                // ensure the patch properly updated the in-memory publish time
+                patchSuccessful = publishTime.getTime() != updatedPublishTime.getTime();
+            }
+
+            // if the patch failed to apply, force a full manifest refresh
+            if (!patchSuccessful) {
+                logger.debug('Patch provided is invalid, performing full manifest refresh');
+                refreshManifest(true);
+                return;
+            }
```

Comment:
I rearranged this a bit to better capture the multiple paths that can force a full refresh, with the full refresh stopping the remainder of the update logic here.

The one question is around the publish time check, I performed it within the updater as it seemed like a more appropriate place seeing as the updater is the time aware component in the process. However, it could also be easily folded into the `applyPatchToManifest` method instead and have that return the patch successful signal as a boolean. If you would prefer to have the adapter method cover it just let me know and I'll make the change.

Reply:
I agree that this is the right place to check